vertx-web/src/main/java/io/vertx/ext/web/handler/impl/OAuth2AuthHandlerImpl.java

```diff
@@ -109,7 +112,11 @@ public void parseCredentials(RoutingContext context, Handler<AsyncResult<JsonObj
       if (token == null) {
         // redirect request to the oauth2 server as we know nothing about this request
         if (callback == null) {
-          handler.handle(Future.failedFuture("callback route is not configured."));
+          // it's a failure both cases but the cause is not the same
+          handler.handle(Future.failedFuture(
```

Comment:
If token == null then a HTTP 401 should be returned in the bearer-only case, and a HTTP 500 instead in the full oauth2 case. I would propose to make a change like e.g.:
...
    if (bearerOnly) {
        handler.handle(Future.failedFuture(new HttpStatusException(HttpURLConnection.HTTP_UNAUTHORIZED,
                            "bearer-only access type: token is null")));
    } else {
        handler.handle(Future.failedFuture("callback route is not configured."));
    }
...

Reply:
I'm inclined in undoing this line change, because the 2nd argument of `parseAuthorization` tells that method if this header is optional or not. When the handler is working in `bearerOnly` the value becomes `false` which means mandatory, and in that case, the token will never be `null`.

If the header is missing or is null, then a 401 is already raised.

Reply:
Indeed cfr. RFC 6750 chapter 3.1 if the request is missing a required parameter - e.g. the token in this case - then a HTTP 400 (Bad Request) must be returned which is exactly what the parseAuthorization method is doing. You're right to say there is actually no request that would lead to a token == null situation after a successful parseAuthorization in this method, it would be an internal programmatic error. Maybe good to put a comment?